README.md
universal-deep-strict-equal
================================

A port of [Node v6's internal _deepEqual function](https://github.com/nodejs/node/blob/v6.1.0/lib/assert.js#L146) in a universal style.

[![Build Status][travis-image]][travis-url]
[![NPM version][npm-image]][npm-url]
[![License][license-image]][license-url]

[![Sauce Test Status][saucelabs-image]][saucelabs-url]

*Issues and improvements should be done in [Node.js](https://github.com/nodejs/node/issues) first.*


API
---------------------------------------

`deepEqual(actual, expected, [strict])`


USAGE
---------------------------------------

```javascript
var deepEqual = require('universal-deep-strict-equal');

deepEqual({a:1}, {a:'1'}); // => true
deepEqual({a:1}, {a:'1'}, false); // => true
deepEqual({a:1}, {a:'1'}, true); // => false
```


INSTALL
---------------------------------------

```
npm install universal-deep-strict-equal
```


AUTHOR
---------------------------------------
* [Takuto Wada](https://github.com/twada)


CONTRIBUTORS
---------------------------------------
* [azu](https://github.com/azu)


LICENSE
---------------------------------------
Licensed under the [MIT](http://twada.mit-license.org/) license.


[travis-url]: https://travis-ci.org/twada/universal-deep-strict-equal
[travis-image]: https://secure.travis-ci.org/twada/universal-deep-strict-equal.svg?branch=master

[npm-url]: https://npmjs.org/package/universal-deep-strict-equal
[npm-image]: https://badge.fury.io/js/universal-deep-strict-equal.svg

[license-url]: http://twada.mit-license.org/
[license-image]: https://img.shields.io/badge/license-MIT-brightgreen.svg

[saucelabs-url]: https://saucelabs.com/u/udse
[saucelabs-image]: https://saucelabs.com/browser-matrix/udse.svg


CHANGELOG.md
### [1.2.1](https://github.com/twada/universal-deep-strict-equal/releases/tag/v1.2.1) (2016-05-06)


#### Bug Fixes

  * [Dealing with old IE that doesn't have Arguments](https://github.com/twada/universal-deep-strict-equal/pull/7)


## [1.2.0](https://github.com/twada/universal-deep-strict-equal/releases/tag/v1.2.0) (2016-05-06)


#### Features

  * [Allow circular references](https://github.com/twada/universal-deep-strict-equal/pull/4) by [@azu](https://github.com/azu)


## [1.1.0](https://github.com/twada/universal-deep-strict-equal/releases/tag/v1.1.0) (2016-05-06)


#### Features

  * [Support old browsers](https://github.com/twada/universal-deep-strict-equal/pull/2)


## [1.0.0](https://github.com/twada/universal-deep-strict-equal/releases/tag/v1.0.0) (2016-05-04)


#### Features

  * the first stable release


## [0.1.0](https://github.com/twada/universal-deep-strict-equal/releases/tag/v0.1.0) (2016-05-03)


#### Features

  * initial release


test/test-deep-equal.js
// port of https://github.com/nodejs/node/blob/v6.1.0/test/parallel/test-assert.js#L62-L163
//     and https://github.com/nodejs/node/blob/v6.1.0/test/parallel/test-assert.js#L384-L410
'use strict';

var _deepEqual = require('..');
var assert = require('assert');
var keys = Object.keys || require('object-keys');

function deepEqual (actual, expected) {
    return _deepEqual(actual, expected, false);
}

describe('deepEqual', function () {

it('7.2', function () {
assert(deepEqual(new Date(2000, 3, 14), new Date(2000, 3, 14)), 'deepEqual(new Date(2000, 3, 14), new Date(2000, 3, 14))');
assert(!deepEqual(new Date(), new Date(2000, 3, 14)), 'deepEqual(new Date(), new Date(2000, 3, 14))');
});


it('7.3', function () {
assert(deepEqual(/a/, /a/));
assert(deepEqual(/a/g, /a/g));
assert(deepEqual(/a/i, /a/i));
assert(deepEqual(/a/m, /a/m));
assert(deepEqual(/a/igm, /a/igm));
assert(!deepEqual(/ab/, /a/));
assert(!deepEqual(/a/g, /a/));
assert(!deepEqual(/a/i, /a/));
assert(!deepEqual(/a/m, /a/));
assert(!deepEqual(/a/igm, /a/im));

var re1 = /a/;
re1.lastIndex = 3;
assert(!deepEqual(re1, /a/));
});


it('7.4', function () {
assert(deepEqual(4, '4'), 'deepEqual(4, \'4\')');
assert(deepEqual(true, 1), 'deepEqual(true, 1)');
assert(!deepEqual(4, '5'), 'deepEqual( 4, \'5\')');
});


it('7.5', function () {
// having the same number of owned properties && the same set of keys
assert(deepEqual({a: 4}, {a: 4}));
assert(deepEqual({a: 4, b: '2'}, {a: 4, b: '2'}));
assert(deepEqual([4], ['4']));
assert(!deepEqual({a: 4}, {a: 4, b: true}));
assert(deepEqual(['a'], {0: 'a'}));
//(although not necessarily the same order),
assert(deepEqual({a: 4, b: '1'}, {b: '1', a: 4}));
var a1 = [1, 2, 3];
var a2 = [1, 2, 3];
a1.a = 'test';
a1.b = true;
a2.b = true;
a2.a = 'test';
assert(!deepEqual(keys(a1), keys(a2)));
assert(deepEqual(a1, a2));
});

it('having an identical prototype property', function () {
var nbRoot = {
  toString: function() { return this.first + ' ' + this.last; }
};

function nameBuilder(first, last) {
  this.first = first;
  this.last = last;
  return this;
}
nameBuilder.prototype = nbRoot;

function nameBuilder2(first, last) {
  this.first = first;
  this.last = last;
  return this;
}
nameBuilder2.prototype = nbRoot;

var nb1 = new nameBuilder('Ryan', 'Dahl');
var nb2 = new nameBuilder2('Ryan', 'Dahl');

assert(deepEqual(nb1, nb2));

nameBuilder2.prototype = Object;
nb2 = new nameBuilder2('Ryan', 'Dahl');
assert(deepEqual(nb1, nb2));
});

it('primitives and object', function () {
assert(!deepEqual(null, {}));
assert(!deepEqual(undefined, {}));
assert(!deepEqual('a', ['a']));
assert(!deepEqual('a', {0: 'a'}));
assert(!deepEqual(1, {}));
assert(!deepEqual(true, {}));
if (typeof Symbol !== 'undefined') {
  assert(!deepEqual(Symbol(), {}));
}
});

// https://github.com/nodejs/node/issues/6416
it("Make sure circular refs don't throw", function(){
var b = {};
b.b = b;

var c = {};
c.b = c;

assert(deepEqual(b, c));
assert(deepEqual(b, c));

var d = {};
d.a = 1;
d.b = d;

var e = {};
e.a = 1;
e.b = e.a;

assert(!deepEqual(d, e));
assert(!deepEqual(d, e));
});

it("GH-7178. Ensure reflexivity of deepEqual with `arguments` objects", function(){
var args = (function() { return arguments; })();
assert(!deepEqual([], args));
assert(!deepEqual(args, []));
});

  describe('primitive wrappers and object', function () {
    it('String and array', function () {
      if (new String('a')['0'] === 'a') {
        assert(deepEqual(new String('a'), ['a']));
      } else {
        assert(!deepEqual(new String('a'), ['a']));
      }
    });
    it('String and object', function () {
      if (new String('a')['0'] === 'a') {
        assert(deepEqual(new String('a'), {0: 'a'}));
      } else {
        assert(!deepEqual(new String('a'), {0: 'a'}));
      }
    });
    it('Number', function () {
      assert(deepEqual(new Number(1), {}));
    });
    it('Boolean', function () {
      assert(deepEqual(new Boolean(true), {}));
    });
  });

});


test/test-deep-strict-equal.js
// port of https://github.com/nodejs/node/blob/v6.1.0/test/parallel/test-assert.js#L165-L280
'use strict';

var _deepEqual = require('..');
var assert = require('assert');

function deepStrictEqual (actual, expected) {
    return _deepEqual(actual, expected, true);
}

describe('deepStrictEqual', function () {

it('7.2', function () {
assert(deepStrictEqual(new Date(2000, 3, 14), new Date(2000, 3, 14)), 'deepStrictEqual(new Date(2000, 3, 14), new Date(2000, 3, 14))');

assert(!deepStrictEqual(new Date(), new Date(2000, 3, 14)), 'deepStrictEqual(new Date(), new Date(2000, 3, 14))');
});


it('7.3 - strict', function () {
assert(deepStrictEqual(/a/, /a/));
assert(deepStrictEqual(/a/g, /a/g));
assert(deepStrictEqual(/a/i, /a/i));
assert(deepStrictEqual(/a/m, /a/m));
assert(deepStrictEqual(/a/igm, /a/igm));
assert(!deepStrictEqual(/ab/, /a/));
assert(!deepStrictEqual(/a/g, /a/));
assert(!deepStrictEqual(/a/i, /a/));
assert(!deepStrictEqual(/a/m, /a/));
assert(!deepStrictEqual(/a/igm, /a/im));

var re1 = /a/;
re1.lastIndex = 3;
assert(!deepStrictEqual(re1, /a/));

});


it('7.4 - strict', function () {
assert(!deepStrictEqual(4, '4'), 'deepStrictEqual(4, \'4\')');
assert(!deepStrictEqual(true, 1),'deepStrictEqual(true, 1)');
assert(!deepStrictEqual(4, '5'), 'deepStrictEqual(4, \'5\')');
});


it('7.5 - strict', function () {
// having the same number of owned properties && the same set of keys
assert(deepStrictEqual({a: 4}, {a: 4}));
assert(deepStrictEqual({a: 4, b: '2'}, {a: 4, b: '2'}));
assert(!deepStrictEqual([4], ['4']));
assert(!deepStrictEqual({a: 4}, {a: 4, b: true}));
assert(!deepStrictEqual(['a'], {0: 'a'}));
//(although not necessarily the same order),
assert(deepStrictEqual({a: 4, b: '1'}, {b: '1', a: 4}));

assert(!deepStrictEqual([0, 1, 2, 'a', 'b'], [0, 1, 2, 'b', 'a']));

var a1 = [1, 2, 3];
var a2 = [1, 2, 3];
a1.a = 'test';
a1.b = true;
a2.b = true;
a2.a = 'test';

assert(deepStrictEqual(a1, a2));
});


it('Prototype check', function () {
function Constructor1(first, last) {
  this.first = first;
  this.last = last;
}

function Constructor2(first, last) {
  this.first = first;
  this.last = last;
}

var obj1 = new Constructor1('Ryan', 'Dahl');
var obj2 = new Constructor2('Ryan', 'Dahl');

assert(!deepStrictEqual(obj1, obj2));

Constructor2.prototype = Constructor1.prototype;
obj2 = new Constructor2('Ryan', 'Dahl');

assert(deepStrictEqual(obj1, obj2));
});


it('primitives', function () {
assert(!deepStrictEqual(4, '4'));
assert(!deepStrictEqual(true, 1));
if (typeof Symbol !== 'undefined') {
  assert(!deepStrictEqual(Symbol(), Symbol()));
  var s = Symbol();
  assert(deepStrictEqual(s, s));
}
});


it('primitives and object', function () {
assert(!deepStrictEqual(null, {}));
assert(!deepStrictEqual(undefined, {}));
assert(!deepStrictEqual('a', ['a']));
assert(!deepStrictEqual('a', {0: 'a'}));
assert(!deepStrictEqual(1, {}));
assert(!deepStrictEqual(true, {}));
if (typeof Symbol !== 'undefined') {
  assert(!deepStrictEqual(Symbol(), {}));
}
});


it('primitive wrappers and object', function () {
assert(!deepStrictEqual(new String('a'), ['a']));
assert(!deepStrictEqual(new String('a'), {0: 'a'}));
assert(!deepStrictEqual(new Number(1), {}));
assert(!deepStrictEqual(new Boolean(true), {}));
});

});

// https://github.com/nodejs/node/issues/6416
it("Make sure circular refs don't throw", function(){
var b = {};
b.b = b;

var c = {};
c.b = c;

assert(deepStrictEqual(b, c));

var d = {};
d.a = 1;
d.b = d;

var e = {};
e.a = 1;
e.b = e.a;

assert(!deepStrictEqual(d, e));
});


test/test-assert-typedarray-deepequal.js
// port of https://github.com/nodejs/node/blob/v6.1.0/test/parallel/test-assert-typedarray-deepequal.js
'use strict';

var _deepEqual = require('..');
var assert = require('assert');

function deepEqual (actual, expected) {
    return _deepEqual(actual, expected, false);
}

function supportsTypedArray() {
    try {
        var a1 = new Uint8ClampedArray(1e5);
        var a2 = new Float64Array(1e5);
        return (a1 && a2);
    } catch (e) {
        return false;
    }
}

if (supportsTypedArray()) {

var equalArrayPairs = [
  [new Uint8Array(1e5), new Uint8Array(1e5)],
  [new Uint16Array(1e5), new Uint16Array(1e5)],
  [new Uint32Array(1e5), new Uint32Array(1e5)],
  [new Uint8ClampedArray(1e5), new Uint8ClampedArray(1e5)],
  [new Int8Array(1e5), new Int8Array(1e5)],
  [new Int16Array(1e5), new Int16Array(1e5)],
  [new Int32Array(1e5), new Int32Array(1e5)],
  [new Float32Array(1e5), new Float32Array(1e5)],
  [new Float64Array(1e5), new Float64Array(1e5)],
  [new Int16Array(256), new Uint16Array(256)],
  [new Int16Array([256]), new Uint16Array([256])],
  [new Float32Array([+0.0]), new Float32Array([-0.0])],
  [new Float64Array([+0.0]), new Float32Array([-0.0])],
  [new Float64Array([+0.0]), new Float64Array([-0.0])]
];

var notEqualArrayPairs = [
  [new Uint8Array(2), new Uint8Array(3)],
  [new Uint8Array([1, 2, 3]), new Uint8Array([4, 5, 6])],
  [new Uint8ClampedArray([300, 2, 3]), new Uint8Array([300, 2, 3])],
  [new Uint16Array([2]), new Uint16Array([3])],
  [new Uint16Array([0]), new Uint16Array([256])],
  [new Int16Array([0]), new Uint16Array([256])],
  [new Int16Array([-256]), new Uint16Array([0xff00])], // same bits
  [new Int32Array([-256]), new Uint32Array([0xffffff00])], // ditto
  [new Float32Array([0.1]), new Float32Array([0.0])],
  [new Float64Array([0.1]), new Float64Array([0.0])]
];

  describe('TypedArray deepEqual', function () {
    equalArrayPairs.forEach(function (arrayPair, idx) {
      it('equalArrayPairs - ' + idx, function () {
        assert(deepEqual(arrayPair[0], arrayPair[1]));
      });
    });
    notEqualArrayPairs.forEach(function (arrayPair, idx) {
      it('notEqualArrayPairs - ' + idx, function () {
        assert(!deepEqual(arrayPair[0], arrayPair[1]));
      });
    });
  });
} else {
  it('SKIP -- this platform does not support TypedArray', function () {
    assert(true, 'skipped');
  });
}


index.js
// port of https://github.com/nodejs/node/blob/v6.1.0/lib/assert.js#L146-L254

// http://wiki.commonjs.org/wiki/Unit_Testing/1.0
//
// Originally from narwhal.js (http://narwhaljs.org)
// Copyright (c) 2009 Thomas Robinson <280north.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the 'Software'), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

'use strict';

var Buffer = require('buffer').Buffer;
var compare = Buffer.compare;
var indexOf = require('indexof');
var filter = require('array-filter');
var pSlice = Array.prototype.slice;
var getPrototypeOf = Object.getPrototypeOf || function(obj) {
  return obj.__proto__ || (
    obj.constructor
      ? obj.constructor.prototype
      : Object.prototype
  );
};
function isEnumerable(obj, key) {
  return Object.prototype.propertyIsEnumerable.call(obj, key);
};
function pToString(obj) {
  return Object.prototype.toString.call(obj);
};
function isPrimitive(arg) {
  return arg === null ||
         typeof arg === 'boolean' ||
         typeof arg === 'number' ||
         typeof arg === 'string' ||
         typeof arg === 'symbol' ||  // ES6 symbol
         typeof arg === 'undefined';
}
function isObject(arg) {
  return typeof arg === 'object' && arg !== null;
}
function isDate(d) {
  return isObject(d) && pToString(d) === '[object Date]';
}
function isRegExp(re) {
  return isObject(re) && pToString(re) === '[object RegExp]';
}
var isArguments = (function () {
  function isArg(obj) {
    return isObject(obj) && pToString(obj) == '[object Arguments]';
  }
  // dealing with old IEs (There's no Arguments type)
  if (!isArg(arguments)) {
    return function(obj) {
      return isObject(obj) &&
        typeof obj.length === 'number' &&
        obj.length >= 0 &&
        pToString(obj) !== '[object Array]' &&
        pToString(obj.callee) === '[object Function]';
    };
  } else {
    return isArg;
  }
})();
function fromBufferSupport() {
  try {
    return typeof Buffer.from === 'function' && !!Buffer.from([0x62,0x75,0x66,0x66,0x65,0x72]);
  } catch (e) {
    // Buffer.from under Node 4.x causes `TypeError: this is not a typed array.`
    return false;
  }
}
var toBuffer = (function () {
  // check whether Buffer constructor accepts ArrayBuffer or not
  function isBufferConstructorAcceptsArrayBuffer() {
    try {
      return typeof Uint8Array === 'function' && (new Buffer(new Uint8Array([1]).buffer)[0] === 1);
    } catch (e) {
      return false;
    }
  }
  if (isBufferConstructorAcceptsArrayBuffer()) {
    // Node 4.x
    return function (ab) {
      return new Buffer(ab);
    };
  } else {
    // Node 0.10.x and 0.12.x
    return function (ab) {
      var buffer = new Buffer(ab.byteLength);
      var view = new Uint8Array(ab);
      for (var i = 0; i < buffer.length; ++i) {
        buffer[i] = view[i];
      }
      return buffer;
    };
  }
})();
var bufferFrom = fromBufferSupport() ? Buffer.from : toBuffer;
var objectKeys = (function () {
  var OLD_V8_ARRAY_BUFFER_ENUM = ['BYTES_PER_ELEMENT','get','set','slice','subarray','buffer','length','byteOffset','byteLength'];
  var keys = Object.keys || require('object-keys');
  return function objectKeys(obj) {
    // avoid iterating enumerable properties of ArrayBuffer under old V8
    if (isEnumerable(obj, 'buffer') &&
        isEnumerable(obj, 'byteOffset') &&
        isEnumerable(obj, 'byteLength')) {
      return filter(keys(obj), function (k) {
        return indexOf(OLD_V8_ARRAY_BUFFER_ENUM, k) === -1;
      });
    } else {
      return keys(obj);
    }
  };
})();

function _deepEqual(actual, expected, strict, memos) {
  // 7.1. All identical values are equivalent, as determined by ===.
  if (actual === expected) {
    return true;
  } else if (actual instanceof Buffer && expected instanceof Buffer) {
    return compare(actual, expected) === 0;

  // 7.2. If the expected value is a Date object, the actual value is
  // equivalent if it is also a Date object that refers to the same time.
  } else if (isDate(actual) && isDate(expected)) {
    return actual.getTime() === expected.getTime();

  // 7.3 If the expected value is a RegExp object, the actual value is
  // equivalent if it is also a RegExp object with the same source and
  // properties (`global`, `multiline`, `lastIndex`, `ignoreCase`).
  } else if (isRegExp(actual) && isRegExp(expected)) {
    return actual.source === expected.source &&
           actual.global === expected.global &&
           actual.multiline === expected.multiline &&
           actual.lastIndex === expected.lastIndex &&
           actual.ignoreCase === expected.ignoreCase;

  // 7.4. Other pairs that do not both pass typeof value == 'object',
  // equivalence is determined by ==.
  } else if ((actual === null || typeof actual !== 'object') &&
             (expected === null || typeof expected !== 'object')) {
    return strict ? actual === expected : actual == expected;

  // If both values are instances of typed arrays, wrap their underlying
  // ArrayBuffers in a Buffer each to increase performance
  // This optimization requires the arrays to have the same type as checked by
  // Object.prototype.toString (aka pToString). Never perform binary
  // comparisons for Float*Arrays, though, since e.g. +0 === -0 but their
  // bit patterns are not identical.
  } else if (typeof ArrayBuffer === 'function' && typeof ArrayBuffer.isView === 'function' &&
             ArrayBuffer.isView(actual) && ArrayBuffer.isView(expected) &&
             pToString(actual) === pToString(expected) &&
             !(actual instanceof Float32Array ||
               actual instanceof Float64Array)) {
    return compare(bufferFrom(actual.buffer),
                   bufferFrom(expected.buffer)) === 0;

  // 7.5 For all other Object pairs, including Array objects, equivalence is
  // determined by having the same number of owned properties (as verified
  // with Object.prototype.hasOwnProperty.call), the same set of keys
  // (although not necessarily the same order), equivalent values for every
  // corresponding key, and an identical 'prototype' property. Note: this
  // accounts for both named and indexed properties on Arrays.
  } else {
    memos = memos || {actual: [], expected: []};

    var actualIndex = indexOf(memos.actual, actual);
    if (actualIndex !== -1) {
      if (actualIndex === indexOf(memos.expected, expected)) {
        return true;
      }
    }

    memos.actual.push(actual);
    memos.expected.push(expected);

    return objEquiv(actual, expected, strict, memos);
  }
}

function objEquiv(a, b, strict, actualVisitedObjects) {
  if (a === null || a === undefined || b === null || b === undefined)
    return false;
  // if one is a primitive, the other must be same
  if (isPrimitive(a) || isPrimitive(b))
    return a === b;
  if (strict && getPrototypeOf(a) !== getPrototypeOf(b))
    return false;
  var aIsArgs = isArguments(a),
      bIsArgs = isArguments(b);
  if ((aIsArgs && !bIsArgs) || (!aIsArgs && bIsArgs))
    return false;
  if (aIsArgs) {
    a = pSlice.call(a);
    b = pSlice.call(b);
    return _deepEqual(a, b, strict);
  }
  var ka = objectKeys(a),
      kb = objectKeys(b),
      key, i;
  // having the same number of owned properties (keys incorporates
  // hasOwnProperty)
  if (ka.length != kb.length)
    return false;
  //the same set of keys (although not necessarily the same order),
  ka.sort();
  kb.sort();
  //~~~cheap key test
  for (i = ka.length - 1; i >= 0; i--) {
    if (ka[i] != kb[i])
      return false;
  }
  //equivalent values for every corresponding key, and
  //~~~possibly expensive deep test
  for (i = ka.length - 1; i >= 0; i--) {
    key = ka[i];
    if (!_deepEqual(a[key], b[key], strict, actualVisitedObjects))
      return false;
  }
  return true;
}

module.exports = _deepEqual;
